File: Beecrownd/2552-PaoDeQueijo/main.java
import java.util.Scanner;

class Celula {
    int valor;
    Celula dir, esq, sup, inf;

    public Celula(int valor) {
        this.valor = valor;
        this.dir = null;
        this.esq = null;
        this.sup = null;
        this.inf = null;
    }
}

class Matriz {
    private Celula inicio;
    int linha, coluna;

    public Matriz(int linha, int coluna, Scanner scanner) {
        this.linha = linha;
        this.coluna = coluna;
        criarMatriz(linha, coluna, scanner);
    }

    private void criarMatriz(int linha, int coluna, Scanner scanner) {
        Celula[][] mat = new Celula[linha][coluna];

        for (int i = 0; i < linha; i++) {
            for (int j = 0; j < coluna; j++) {
                int x = scanner.nextInt();
                mat[i][j] = new Celula(x);
            }
        }

        for (int i = 0; i < linha; i++) {
            for (int j = 0; j < coluna; j++) {
                if (i > 0) mat[i][j].sup = mat[i - 1][j];
                if (i < linha - 1) mat[i][j].inf = mat[i + 1][j];
                if (j > 0) mat[i][j].esq = mat[i][j - 1];
                if (j < coluna - 1) mat[i][j].dir = mat[i][j + 1];
            }
        }

        inicio = mat[0][0];
    }

    public void paoDeQueijo() {
        for (Celula linhaAtual = inicio; linhaAtual != null; linhaAtual = linhaAtual.inf) {
            for (Celula atual = linhaAtual; atual != null; atual = atual.dir) {
                if (atual.valor == 1) {
                    System.out.print("9");
                } else {
                    int cont = 0;
                    if (atual.sup != null && atual.sup.valor == 1) cont++;
                    if (atual.inf != null && atual.inf.valor == 1) cont++;
                    if (atual.esq != null && atual.esq.valor == 1) cont++;
                    if (atual.dir != null && atual.dir.valor == 1) cont++;
                    System.out.print(cont);
                }
            }
            System.out.println();
        }
    }
}

public class main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (scanner.hasNextInt()) {
            int n = scanner.nextInt();
            int m = scanner.nextInt();

            Matriz matriz = new Matriz(n, m, scanner);
            matriz.paoDeQueijo();
        }

        scanner.close();
    }
}

File: Beecrownd/2552-PaoDeQueijo/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula{
    int valor; 
    struct Celula *dir;
    struct Celula *esq;
    struct Celula *sup;
    struct Celula *inf;

}Celula;

typedef struct Matriz{
    Celula ***dados;
    Celula *inicio;
    int linha;
    int coluna;

}Matriz;

Matriz *criarMatriz(int linha, int coluna){
    Matriz *matriz = (Matriz*)malloc(sizeof(Matriz));
    matriz-> linha = linha;
    matriz->coluna = coluna;

    matriz->dados = (Celula***)malloc(linha * sizeof(Celula**));

    for(int i = 0; i < linha; i++){
        matriz->dados[i] = (Celula**)malloc(coluna * sizeof(Celula*));
        for(int j = 0; j < coluna; j++){
            matriz->dados[i][j] = (Celula*)malloc(sizeof(Celula));
            matriz->dados[i][j]->valor = 0;
            matriz->dados[i][j]->dir = NULL;
            matriz->dados[i][j]->esq = NULL;
            matriz->dados[i][j]->sup = NULL;
            matriz->dados[i][j]->inf = NULL;
        }
    }
    
    for(int i = 0; i < linha; i++){
        for(int j = 0; j < coluna; j++){
            if(i > 0){
                matriz->dados[i][j]->sup =  matriz->dados[i - 1][j];
            }
            if(i < linha -1){
                matriz->dados[i][j]->inf =  matriz->dados[i + 1][j];
            }
            if(j > 0){
                matriz->dados[i][j]->esq =  matriz->dados[i][j - 1];
            }
            if(j < coluna - 1){
                matriz->dados[i][j]->dir =  matriz->dados[i][j + 1];
            }
        }
    }

    matriz->inicio = matriz->dados[0][0];

    return matriz;

}

void preencher(Matriz *matriz){
    for(int i = 0; i < matriz->linha; i++){
        for(int j = 0; j < matriz->coluna; j++){
            scanf("%d", &matriz->dados[i][j]->valor);
        }
    }
}


void paoqueijo(Matriz *matriz){
    
    for(int i = 0; i < matriz->linha; i++){
        for(int j = 0; j < matriz->coluna; j++){
            if(matriz->dados[i][j]->valor == 1){
                printf("9");
            }else{
                int cont = 0;
                if(matriz->dados[i][j]-> dir != NULL && matriz->dados[i][j]->dir->valor == 1){cont++;}
                if(matriz->dados[i][j]-> esq != NULL && matriz->dados[i][j]->esq->valor == 1){cont++;}
                if(matriz->dados[i][j]-> sup != NULL && matriz->dados[i][j]->sup->valor == 1){cont++;}
                if(matriz->dados[i][j]-> inf != NULL && matriz->dados[i][j]->inf->valor == 1){cont++;}
                printf("%d", cont);
            }

        }

        printf("\n");
    }
}

/*void imprimir(Matriz *matriz){
    for(int i = 0; i < matriz->linha; i++){
        for(int j = 0; j < matriz->coluna; j++){
            printf("%d ", matriz->dados[i][j]->valor);
        }

        printf("\n");
    }

}*/

int main(){
    int linha,coluna;

    while(scanf("%d %d", &linha, &coluna) != EOF ){
        Matriz *matriz = criarMatriz(linha,coluna);

        preencher(matriz);
        paoqueijo(matriz);
    }
}
